docs/architecture/adr-110-e2e-abci++.md

```diff
@@ -0,0 +1,267 @@
+`GetExecutionString()` function. This function receives a list of `abci.Request` and generates a string where every request the grammar covers 
+will be replaced with a corresponding terminal. For example, `abci.Request_PrepareProposal` is replaced with `<PrepareProposal>`. If the request is not covered 
+by the grammar, it will be ignored. 
+- Checks if the resulting string with terminals respects the grammar. This 
+logic is implemented inside the `VerifyExecution` function. 
+
+```go
+func (g *GrammarChecker) VerifyExecution(execution string) (bool, error) {
+	lexer := lexer.New([]rune(execution))
+	_, errs := parser.Parse(lexer)
+	if len(errs) > 0 {
+		err := g.combineParseErrors(execution, errs, g.cfg.MaxNumberOfErrorsToShow)
+		return false, err
+	}
+	return true, nil
+}
+```
+This function is the only function that uses auto-generated parser and 
+lexer. It returns true if the execution is valid. Otherwise, it returns an 
+error composed of parser errors and some additional information 
+we added. An example of an error produced by `VerifyExecution`
+is the following:
+
+```
+Parser failed, number of errors is 11
+    ***Error 0***
+    Parse Error: ProcessProposal : ∙<ProcessProposal>  I[11]=<Commit> (193,201) <Commit> at line 3 col 19
+    Expected one of: [<FinalizeBlock>,<PrepareProposal>,<ProcessProposal>]
+    Execution:
+    1:<InitChain> <ProcessProposal> <ProcessProposal> <ProcessProposal> <ProcessProposal> 
+    2:<ProcessProposal> <ProcessProposal> <ProcessProposal> <ProcessProposal> <ProcessProposal>
+	3:<ProcessProposal> <Commit>
+```
+The parse error shown above represents an error that happened at the grammar slot
+`ProcessProposal : ∙<ProcessProposal>`, specifically an error occurs at token 
+`I[11]=<Commit> (193,201) <Commit>` which is 12th token in the whole 
+execution, at line 3, column 19. Instead of `<Commit>` the grammar was 
+expecting `<FinalizeBlock>`,`<PrepareProposal>` or `<ProcessProposal>`.
+In addition, the output shows the lines around the line with an error. 
+Notice here that the parser can return many errors because the parser returns an error at every point at which the parser fails to parse
+a grammar production. Usually, the error of interest is the one that has 
+parsed the largest number of tokens. This is why, at the 
+moment, we are printing the last 10 errors; however, this is part of the configuration and can be changed. 
+
+### Suporting additional ABCI++ requests
+
+Here we present all the steps we need to do if we want to support other 
+ABCI++ requests in the future: 
+
+- The application needs to log the new request in the same way as we do now.
+- We should include the new request to the grammar and generate a new parser and lexer.  
+- We should modify `getRequestTerminal()` method inside `test/e2e/pkg/grammar/checker.go` to return a grammar terminal for the new ABCI++ request. 
+
+
+## Status
+
+Implemented.
+
+To-do list:
+- integrating the generation of parser/lexer into the codebase.
```

Comment:
How are you planning to do this? (I think it should be explained in this ADR)

My advice is to be as similar as possible to other generated code we have (e.g., protos):
* you have "make proto" to generate the code (so you could have "make grammar")
* the autogenerated code is committed to the repo (so people don't need to do "make grammar" if they don't change it)
* there is a CI workflow that checks for every PR that the author didn't forget to run "make proto" (so we could think of something similar here)